### ISSUE TITLE
Document how to use a local Docker repository with Pachyderm

### ISSUE THREAD

**Issue by bpeabody** [pachyderm/pachyderm#4164]
I appreciate that the Pachyderm tutorial allows you to do everything on your local machine. However, once you are to a point where you want to experiment with your own images, you may find yourself in a bind unless you are (a) willing to put them in Docker Hub or (b) have a secured registry like AWS ECR. Even then, there is something nice about being able to work disconnected.

Thankfully, it's pretty easy to set up a local Docker registry for this purpose. I used a tool called [trow](https://github.com/ContainerSolutions/trow), following the directions in this [blog post](https://blog.container-solutions.com/installing-a-registry-on-kubernetes-quickstart)

NOTE: Do **NOT** say “y” to the prompt where trow asks if you want it to install itself as a validating web hook. If you do so it will not allow access to Pachyderm images from your cluster. If you happen to make this mistake, the only way I found to rectify was to rebuild the cluster.

First, install trow:

```
$ git clone https://github.com/ContainerSolutions/trow.git
$ cd trow
$ ./install.sh
```

Once this is complete  you must restart your Docker daemon. Then you will be able to use your images locally. After you push the image to trow’s registry:

```
$ docker tag myimage trow.kube-public:31000/myimage
$ docker push trow.kube-public:31000/myimage
```

You can reference it (e.g., as `trow.kube-public:31000/myimage`) in your Pachyderm workflows.

**Comment by pappasilenus** [comment]
Not sure the priority on this, @svekars, given the newness of Trow as a registry. Also, based on internal discussions: There may not be a good reason to use a registry for local development, as you can just build the images directly onto the docker host which is easier to setup, has fewer moving parts so is more reliable, and also more performant.

Feel free to close if we're going to document other mechanisms for accomplishing the same goal.

**Comment by brock-eai** [comment]
I would greatly prefer a solution for local development that has fewer moving parts, is more reliable, and more performant; Trow was just the first thing that I came up with that worked. Maybe you could instead document that approach?

**Comment by svekars** [comment]
There is a way to do this without installing a registry. With minikube,you can transfer your local image to a minikube VM by running this command:

```
docker save <image:version> | (\
  eval $(minikube docker-env)
  docker load
)```

**Comment by mindthevirt** [comment]
This is no longer needed if you use Docker Desktop with Kubernetes. After building your docker image locally, you can just reference the newly build image and tag in the pipeline and it will work without pushing it to docker hub or any other registry.

**Comment by srajwomply** [comment]
Can you give an example on how to reference the local docker image?

**Comment by echohack** [comment]
@srajwomply Hey Sundesh. This issue is closed, and we usually answer questions like this over in our Slack community: https://www.pachyderm.com/slack/

I think you're asking "How can you use local docker images for your Pachyderm pipeline?"

You can reference your local Docker env directly with:

```
eval $(minikube docker-env)
```

You can read more about minikube docker-env here: https://minikube.sigs.k8s.io/docs/commands/docker-env/

Thanks and have a great day!

PS:

You could also use local images in your Pachyderm *deployment* (this is a different issue entirely):

The way that you can reference a local docker image is to modify your deployment.json. You can do this with:

```
pachctl deploy local --dry-run > deployment.json
```

and modify the docker images to reference your local images.
